lms/filter-irrelevant-stripe-invoices

anathomical opened the pull request:
## WHAT
Filter out invoices that have been refunded or are for $0
## WHY
This removes those items from the daily emails for Sales-Ops since we don't expect anyone to make Subscriptions for those
## HOW
Add a couple of additional filter conditions when determining which invoices to flag when not connected to a Subscription

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  Yes
Have you deployed to Staging? | Deploying now
Self-Review: Have you done an initial self-review of the code below on Github? | Yes
Spec Review: Have you reviewed the spec and ensured this meets requirements and/or matches design mockups? | N/A

anathomical committed:
Filter out invoices that have been refunded or are for $0

Since we don't expect anyone to make Subscriptions for those

eadams17 commented on services/QuillLMS/app/workers/identify_stripe_invoices_without_subscriptions_worker.rb:
```diff
@@ -29,14 +29,35 @@ def perform
   end
 
+  private def invoice_ids_with_subscriptions
+    @invoice_ids_with_subscrirptions ||= Subscription.where(stripe_invoice_id: stripe_invoice_ids).pluck(:stripe_invoice_id)
```
looks like you have a typo here

anathomical replied:
It's important to cache your subscrirptions!

eadams17 reviewed (approved):
LGTM! just noticed one typo

brendanshean commented on services/QuillLMS/app/workers/identify_stripe_invoices_without_subscriptions_worker.rb:
```diff
@@ -29,14 +29,35 @@ def perform
+    charge.amount == charge.amount_refunded
+  end
+
+  private def non_zero_amount?(invoice)
```
I would rename `positive_amount?` since non_zero implies negative is possible.

brendanshean replied:
I would rename `positive_amount?` since non_zero implies negative is possible.

brendanshean commented on services/QuillLMS/app/workers/identify_stripe_invoices_without_subscriptions_worker.rb:
```diff
@@ -29,14 +29,35 @@ def perform
+
   private def stripe_invoices
-    [].tap do |stripe_invoices|
+    @stripe_invoices ||= [].tap do |stripe_invoices|
```
For clarity, I would use a different variable name for the tap block variable (e.g. `invoices`) than the method name

brendanshean commented on services/QuillLMS/app/workers/identify_stripe_invoices_without_subscriptions_worker.rb:
```diff
@@ -29,14 +29,35 @@ def perform
       RELEVANT_INVOICE_STATUSES.include?(invoice.status) &&
-        !Subscription.exists?(stripe_invoice_id: invoice.id)
+        !invoice_ids_with_subscriptions.include?(invoice.id) &&
+        !invoice_refunded?(invoice) &&
```
What about switching the order of `non_zero_amount?` and `invoice_refunded?`.

The reason being that if `invoice_refunded?` fails it will short circuit the cycle and skip the API call to `Stripe::Charge.retrieve`

brendanshean reviewed (approved):
Just a few refactoring comments but otherwise, looks good to me.

anathomical committed:
Do some clean up based on PR feedback:

- fix a typo
- re-order relevance conditions so the most expensive condition is checked last
- do some name clean up to reduce potential confusion
